### DISCUSSION TITLE
How to send data as client when connected

### DISCUSSION THREAD

**Discussion by T02my** [sirn-se/websocket-php#82] [Q&A]
Hello, I'm sorry for asking this, I'm still rookie with PHP and this technology of Web Sockets.

I want to:

1) Connect to server as client.
2) Listen to server, when server sends info -> do something
3) When something happens on my client side, I somehow find the existing connection with server and send new info.

1 and 2 I know how to do. But third one I don't know how to and dunno is it possible at all.

For example,
I established connection, listen to what server sends, and in one time I want to use the same connection to send data back to server without establishing new connection. For now how I use it:

1) 1 php file listens to server orders,
2) second php file creates new connection when I need to add new data to server, sends data and closes connection.

But because of this, I think I use websocket wrong.

**Comment by sirn-se** [comment]
There are various options, depending on what client is doing (that is not a reaction to incoming messages from server).

If the action is limited and not triggered by external input, you can use the [onTick listener](https://github.com/sirn-se/websocket-php/blob/v3.2-main/docs/Listener.md#coroutine---the-tick-listener). This callback will run periodically, independent on incoming messages. But it is blocking server read, so you do not want to run loops or too complex code here.

If you need a more flexible setup, I would recommend using a queue as an intermediate between WebSocket client and the internal code (running as separate scripts). When the internal script need to push to server, it writes message to the queue. Then the WebSocket script can use onTick callback to read the queue and push to server. There are advanced asynchronous queue servers such as [RabbitMQ](https://packagist.org/packages/php-amqplib/php-amqplib) but you can also use a custom database to store the queue.

You might also be able to set something up using the PHP [Fibre feature](https://www.php.net/manual/en/language.fibers.php) or an app server like [Swoole](https://wiki.swoole.com/en/#/) which are coroutine implementations, allowing you yo switch code execution context.
